dali/kernels/common/find/find_first_last_gpu.cuh

```diff
@@ -0,0 +1,187 @@
+struct SampleDesc {
+  int64_t *a_ptr, *b_ptr;  // represents (first, last), (begin, end), or (begin, length) depending
+                           // on the OutputProcessor
+  Predicate predicate;
+  const T *in;
+  int64_t len;
+};
+
+/**
+ * @brief Represents first, last (or other range representations) coordinates
+ */
+struct pair_i64 {
+    int64_t a = 0;
+    int64_t b = 0;
+};
+
+/**
+ * @brief First and last position
+ */
+struct first_last {
+  DALI_HOST_DEV DALI_FORCEINLINE pair_i64 operator()(pair_i64 x) const noexcept {
+    return x;
+  }
+};
+
+/**
+ * @brief Begin (inclusive) and End (exclusive) of the region
+ */
+struct begin_end {
+  DALI_HOST_DEV DALI_FORCEINLINE pair_i64 operator()(pair_i64 x) const noexcept {
+    return {x.a, x.b + 1};
+  }
+};
+
+/**
+ * @brief Begin (inclusive) and Length of the region
+ */
+struct begin_length {
+  DALI_HOST_DEV DALI_FORCEINLINE pair_i64 operator()(pair_i64 x) const noexcept {
+    return {x.a, x.b - x.a + 1};
+  }
+};
+
+/**
+ * @brief Extract the position of the first and last position that satisfies a predicate
+ *
+ * @remarks Calculates a double reduction (min, max) in one go
+ *
+ * @tparam T Input data type
+ * @tparam Predicate Predicate that must be satisfied
+ * @tparam OutFormat Optional Transformation to first, last coordinates
+ * @param samples Sample descriptors, including a per-sample predicate
+ * @param format Optional output coordinate transformation
+ */
+template <typename T, typename Predicate, typename OutFormat = first_last>
+__global__ void FindFirstLastImpl(SampleDesc<T, Predicate> *samples, OutFormat format = {}) {
+  int sample_idx = blockIdx.y;
+  auto &sample = samples[sample_idx];
+  const T *input = sample.in;
+  int64_t sample_len = sample.len;
+  auto &predicate = sample.predicate;
+
+  const int64_t blk_size = blockDim.x * blockDim.y;
+  const int64_t grid_size = gridDim.x * blk_size;
+  const int flat_tid = threadIdx.x + threadIdx.y * blockDim.x;
+
+  reductions::min first_reduction;
+  reductions::max last_reduction;
+
+  int64_t first_neutral = first_reduction.template neutral<int64_t>();
+  int64_t last_neutral = last_reduction.template neutral<int64_t>();
+  int64_t first = first_neutral;
+  int64_t last = last_neutral;
+
+  // similar concept as in reduction kernels
+
+  int64_t idx = blockIdx.x * blk_size + flat_tid;
+  for (; idx < sample_len; idx += grid_size) {
+    int64_t tmp_idx = predicate(input[idx]) ? idx : -1;
+    int64_t first_candidate = tmp_idx < 0 ? first_reduction.template neutral<int64_t>() : tmp_idx;
+    int64_t last_candidate = tmp_idx < 0 ? last_reduction.template neutral<int64_t>() : tmp_idx;
+
+    first_reduction(first, first_candidate);
+    last_reduction(last, last_candidate);
+  }
+
+  BlockReduce(first, first_reduction);
+  BlockReduce(last, last_reduction);
+  if (flat_tid == 0) {
+    if (first == first_neutral || last == last_neutral) {
+      *sample.a_ptr = 0;
+      *sample.b_ptr = 0;
```

Comment:
Isn't the {0, 0} a bit misleading for First Last representation?

Reply:
that's right. I think we can add a `neutral` function to the formater